### DISCUSSION TITLE
asyncua.ua.uaerrors._auto.BadNodeIdUnknown

### DISCUSSION THREAD

**Discussion by Ritchie3** [FreeOpcUa/opcua-asyncio#1719] [Q&A]
I'm at a loss here. I don't know why i keep getting this error.

I'm setting up a test server for my script, so I don't need to connect to the machine itself for testin.
I need to read out a value from a variable with:
Node = client.get_node("ns=0; s=HAAS1_79034.Kuka.CellNumber")
print(await Node.get_value())

this gives the error asyncua.ua.uaerrors._auto.BadNodeIdUnknown

when i try the same with 
Node_bJetsonAlive = client.get_node("i=31201")
it works.

this is the server in UA Expert
![image](https://github.com/user-attachments/assets/b7c6e2a3-942e-4f57-96e1-b6432a507d75)

the ns seems correct. Also tried with other ns settings on the server and client.
I don't see any typos in the path.

my client reads out the opc ua tags on the machine (non python server) just fine, but on a local server, and a 2nd remote server, it fails.

anyone have a clue?

**Comment by AndreasHeine** [comment]
sure you use the identifiertype "String" on a "Numeric" NodeId ^^

**Reply by AndreasHeine** [reply]
in order to use something like that "HAAS1_79034.Kuka.CellNumber" you need to create the "BrowsePath" and use the "TranslateBrowsePathToNodeId" service to get "i=31201"!

but please dont use NamespaceIndex 0 it is reserved for normative stuff! you can use NamespaceIndex 1 (Serverownnamespace) or create a own Namespace for the Machine-Instance (Robot) but be aware for Robotics there is already a Standard how to Model Robots in OPC UA!!!

**Reply by Ritchie3** [reply]
Hi Andreas, thanks for the quick reply. 
It seems translate_browsepath_to_nodeid is missing from asyncua. 
I did find it in Python OPC-UA though.

Is there a solution on the server side, by changing the definition of the variables/folders? 

at the moment i'm using an XML file to build the server:


```
<UANodeSet xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" Version="1.02" LastModified="2013-03-06T05:36:44.0862658Z" xmlns="http://opcfoundation.org/UA/2011/03/UANodeSet.xsd">

  <UAObject NodeId="i=31000" BrowseName="HAAS1_79034">
    <Description>folder for Haas1 tags.</Description>
    <References>
      <Reference ReferenceType="HasTypeDefinition">i=61</Reference>
      <Reference ReferenceType="Organizes" IsForward="false">i=85</Reference>
    </References>
  </UAObject>

  <UAObject NodeId="i=31200" BrowseName="Kuka">
    <Description>A custom object node.</Description>
    <References>
      <Reference ReferenceType="HasTypeDefinition">i=58</Reference>
      <Reference ReferenceType="Organizes" IsForward="false">i=31000</Reference>
    </References>
  </UAObject>

  <UAVariable NodeId="i=31201" BrowseName="CellNumber" DataType="UInt32" AccessLevel="3" UserAccessLevel="3"> 
    <References>
      <Reference ReferenceType="HasTypeDefinition">i=68</Reference>
      <Reference ReferenceType="HasProperty" IsForward="false">i=31200</Reference>
    </References>
    <Value>
      <UInt32>79034</UInt32>
    </Value>
  </UAVariable>


</UANodeSet>
```

**Reply by AndreasHeine** [reply]
Client:

https://github.com/FreeOpcUa/opcua-asyncio/blob/1343fa761ce5b2ece79b64041463b86486b81534/asyncua/client/client.py#L958

Server:

its part of isession!

should be: "server.iserver.isession.translate_browsepaths_to_nodeids()"
https://github.com/FreeOpcUa/opcua-asyncio/blob/1343fa761ce5b2ece79b64041463b86486b81534/asyncua/server/internal_session.py#L185

but the translate browsepaths service has some flavour and is not easy to use for beginners if you dont understand the Namespaces concept and the OPC UA Addressspace...

**Reply by AndreasHeine** [reply]
you should use a proper modeling tool e.g SiOME from Siemens! you xml file might be valid XML but it lacks a lot of OPC UA related stuff!!!

e.g.:
https://github.com/AndreasHeine/SampleServer-node-opcua/blob/main/models/opcroboticstestserver.xml

**Reply by Ritchie3** [reply]
Thank you! I'll give it a go on Monday.
